### DISCUSSION TITLE
Time for a new release?

### DISCUSSION THREAD

**Discussion by mattlacey** [knockout/tko#160] [Q&A]
I can see via git tags that the last stamped alpha was Feb 2020 (v4.0.0-alpha9.0) - but the latest on is alpha5h. From the commit history it looks as though there's a been a bunch of useful additions over the last year too.

I'm currently attempting to replace Knockout with TKO in a rather large repository, and one of the things I need to do is replace inline functions in bindings with arrow syntax functions, and while trying to work out where things were broken I noticed a commit from a few months back that helps identify these.

Is there a particular reason a new alpha hasn't been published to npm? Is it something that can be done or is it not feasible at this time?

**Comment by mattlacey** [comment]
Ok, somehow I'd managed to skip over the lines calling out how to install the reference and knockout builds, and of course `npm install @tko/build.reference` has given me the latest tagged version from the repo. That said, might be time for an 11?

**Reply by danieldickison** [reply]
We recently upgraded our site from ko to tko. We found and fixed a few bugs along the way, and not all of those pull requests have been merged yet so we build from source. Here's our "head" branch if you'd like to use it for reference: https://github.com/bandcampdotcom/tko/tree/dev

**Reply by mattlacey** [reply]
Appreciate that! Good to know you folks are using it too :)

**Comment by nmocruz** [comment]
I have also a project that I would like to upgrade, but not while this is
alpha, not sure even if it was beta.
Is really sad to see tko / ko4 struggling to get a usable version.

**Reply by danieldickison** [reply]
Fwiw we've been running our [patched version](https://github.com/bandcampdotcom/tko/tree/dev) of tko site-wide since last fall and it has been totally solid so I think it could arguably be bumped to beta or RC status when the maintainers get to looking at the PRs and releasing it.

**Reply by mattlacey** [reply]
Yeah, I don't think it really qualifies as alpha-quality software, the labels are more tied the fact that not all the ideas on the wishlist have been included so far I feel.

**Comment by nmocruz** [comment]
Well maybe move some wishlist items to future versions instead of keeping
this ob alpha version eternally.

**Reply by mattlacey** [reply]
I don't disagree - this is just my guess as to what's happening as an observer of the project. @brianmhunt do you have any thoughts on how to approach this? Clearly you've been the lead driver for the whole thing, but seems like a bit of reset might be needed around the various goals etc. with the aim of getting more public use? Figure there's a few ko codebases out there looking to make the switch.

**Reply by brianmhunt** [reply]
I think it's due for a beta release, and then we can make more substantial changes to e.g. types, build process, test process.

If all looks good with the latest PR merges I'm ok with a beta version release.

**Comment by brianmhunt** [comment]
Hi All,  Sorry it's been forever since an update, but this is essentially the first real moment I've had off work since August.  On the TKO front I'm very keen to get another release, and the blockers seem "soft" (e.g. build process, testing process).  I'm going to _try_ to carve out some time to get the current PRs reviewed merged in over the next couple weeks, then review whatever is needed for the first beta release.

There are also a _ton_ of Typescript types that I'm keen to get merged in, and I also think it's very important to get automated testing back up, but those are not essential to a release.  One of the major impediments is that every time I have a moment to work on TKO it's sunk entirely into build tools, or getting Sauce Labs or Browserstack working so we have proper cross-browser testing. (We just started with Playwright at work and have had good results on the main browsers so that may be the way forward, but again, not a blocker.)

The one key time-sink I'm concerned about is the build process, mainly determining how to export things in a way so that they can be reasonably consumed by modern and legacy environments, ranging from CommonJS to ESM and import maps in things ranging from Webpack to Vite/Snowpack.  This is unfortunately non-trivial, so I may end up with narrow targets at the start (e.g. ESM-only) and expand later based on demand.

The upside is that we're using TKO in its current alpha state with a large codebase (hundreds of thousands of lines of code, hundreds of view-components) and have had millions of hits in virtually every browser with virtually no issues, with scaling way past what's ordinarily possible with the VirtualDOM i.e. with tens of thousands of DOM elements with O(1) updates / microseconds.

Thanks for your patience, and I hope we can close out a release in a reasonable timeframe.

**Reply by mattlacey** [reply]
Totally understandably Brian - don't mean to put pressure on! Your work on this is very much appreciated!

Have been taking a crack at moving our code base onto it from KO 3.5.1 and it's all been pretty smooth sailing so far. Having to rewrite a few bindings where we've got some overly complicated logic in the templates (using array methods mainly like `indexOf`). Only other thing that's cropped up in a few places is we've got a few instances where null gets thrown into an `observableArray` and that seems to blow up the `foreach` binding - obviously better not to put `null` in there but wasn't sure if this was a conscious design choice or whether I should raise a bug (and propose a fix)?

**Comment by brianmhunt** [comment]
@danieldickison I've merged in all your PRs, and all looks ok to me, tests passing, builds, etc.  Are you able to give it a try and let me know how it works in your case?

**Reply by danieldickison** [reply]
I will take it for a spin. Should be able to verify by tomorrow, if not today.

**Comment by bcdanieldickison** [comment]
@mattlacey in case you missed it, we have a new 4.0.0-beta1.3 release: #175

This is essentially identical to the version that bandcamp.com has been using in production for several months now without issue so I'd consider it very stable. (Though of course there might be areas of the codebase that we don't happen to touch.)

**Comment by mattlacey** [comment]
I had missed that! Will check it out. Saw a few PRs go through. Anyone
had issues with afterRender hooks not firing? That's been one of our
larger pain points moving from ko.

Good work getting a new build out! 

On May 24, 2022, GitHub ***@***.***> wrote:
> @mattlacey <https://github.com/mattlacey> in case you missed it, we
> have a new 4.0.0-beta1.3 release: #175
> <https://github.com/knockout/tko/discussions/175>
>
> This is essentially identical to the version that bandcamp.com has
> been using in production for several months now without issue so I'd
> consider it very stable. (Though of course there might be areas of the
> codebase that we don't happen to touch.)
>

—
Reply to this email directly, view it on GitHub
<https://github.com/knockout/tko/discussions/160#discussioncomment-
2814524>, or unsubscribe <https://github.com/notifications/unsubscribe-
auth/AAJSWW74NPDLSYKDSKSIJG3VLUSBHANCNFSM5LJVD3DA>.
You are receiving this because you were mentioned.Message ID:
***@***.***>
